File: flow/visual/static/visual/canvas/js/graphics/primitives/geometry.js
'use strict';

class Geometry{
    static glyphVertCone(sampling = 10){
        let vert = [];
        let add = 2 * Math.PI / sampling;
        let size = 0.5;

        for (let i = 0; i < 2 * Math.PI; i += add){
            vert.push(0, size * Math.cos(i), size * Math.sin(i));
            vert.push(0, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(size * 2, 0, 0);
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            vert.push(0, size * Math.cos(i), size * Math.sin(i));
            vert.push(0, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(0, 0, 0);
        }

        return vert;
    }

    static glyphNormCone(sampling = 10){
        let vert = [];
        let add = 2 * Math.PI / sampling;
        let a, b;
        let x = vec3.create();
        let y = vec3.create();
        let c = vec3.create();
        let t = vec3.fromValues(2, 0, 0);

        for (let i = 0; i < 2 * Math.PI; i += add){
            a = vec3.fromValues(0, Math.cos(i), Math.sin(i));
            b = vec3.fromValues(0, Math.cos(i + add), Math.sin(i + add));
            vec3.sub(x, a, t);
            vec3.sub(y, b, t);
            vec3.cross(c, x, y);

            for (let k = 0; k < 3; k++){
                vert.push(c[0], c[1], c[2]);
            }
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            for (let k = 0; k < 3; k++){
                vert.push(-1, 0, 0);
            }
        }

        return vert;
    }

    static glyphVertLine(sampling, height = 4){
        let vert = [];
        let add = 2 * Math.PI / sampling;
        let size = 0.5;

        for (let i = 0; i < 2 * Math.PI; i += add){
            vert.push(0, size * Math.cos(i), size * Math.sin(i));
            vert.push(0, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(size * height, size * Math.cos(i), size * Math.sin(i));

            vert.push(0, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(size * height, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(size * height, size * Math.cos(i), size * Math.sin(i));
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            vert.push(0, size * Math.cos(i), size * Math.sin(i));
            vert.push(0, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(0, 0, 0);
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            vert.push(size * height, size * Math.cos(i + add), size * Math.sin(i + add));
            vert.push(size * height, size * Math.cos(i), size * Math.sin(i));
            vert.push(size * height, 0, 0);
        }

        return vert;
    }

    static glyphNormLine(sampling, height = 4){
        let vert = [];
        let add = 2 * Math.PI / sampling;
        let a, b;
        let x = vec3.create();
        let y = vec3.create();
        let c = vec3.create();
        let t = vec3.create();

        for (let i = 0; i < 2 * Math.PI; i += add){
            a = vec3.fromValues(0, Math.cos(i), Math.sin(i));
            b = vec3.fromValues(0, Math.cos(i + add), Math.sin(i + add));
            t = vec3.fromValues(height, Math.cos(i), Math.sin(i));
            vec3.sub(x, a, t);
            vec3.sub(y, b, t);
            vec3.cross(c, x, y);

            for (let k = 0; k < 6; k++){
                vert.push(c[0], c[1], c[2]);
            }
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            for (let k = 0; k < 3; k++){
                vert.push(-1, 0, 0);
            }
        }

        for (let i = 0; i < 2 * Math.PI; i += add){
            for (let k = 0; k < 3; k++){
                vert.push(1, 0, 0);
            }
        }

        return vert;
    }


    static streamVert(sampling, divisions = 10){
        let vert = [];

        for (let i = 0; i < divisions; ++i){
            vert = vert.concat(Geometry.glyphVertLine(sampling, 0));
        }

        return vert;
    } 


    static streamNorm(sampling, divisions = 10){
        let vert = [];

        for (let i = 0; i < divisions; ++i){
            vert = vert.concat(Geometry.glyphNormLine(sampling, 1));
        }

        return vert;
    }


    static streamLocalT(sampling, divisions = 10){
        let vert = [];
        let add = 2 * Math.PI / sampling;

        for (let i = 0; i < divisions; ++i){
            let base = i / divisions;
            let top = (i + 1) / divisions;

            for (let i = 0; i < 2 * Math.PI; i += add){
                vert.push(base, base, top, base, top, top);
            }

            for (let i = 0; i < 2 * Math.PI; i += add){
                vert.push(base, base, base);
            }

            for (let i = 0; i < 2 * Math.PI; i += add){
                vert.push(top, top, top);
            }

        }

        return vert;
    }

    static layerElements(sampling, normal){
        sampling.splice(normal,1);

        let elements = [];
        for (let b = 0; b < sampling[1] - 1; ++b){
            for (let a = 0; a < sampling[0] - 1; ++a){
                elements.push(a + 1 + b * sampling[1], a + b * sampling[1], a + sampling[0] + b * sampling[1]);
                elements.push(a + 1 + b * sampling[1], a + sampling[0] + b * sampling[1], a + 1 + sampling[0] + b * sampling[1]);
            }
        }

        return elements;

    }

    static get unitQuad(){
        return [
            -0.5, -0.5, 0,
            0.5, -0.5, 0,
            0.5, 0.5, 0,
            -0.5, -0.5, 0,
            0.5, 0.5, 0,
            -0.5, 0.5, 0,
        ];
    }

    static get unitQuadTex(){
        return [
            0, 1,
            1, 1,
            1, 0,
            0, 1,
            1, 0,
            0, 0,
        ];
    }
}

File: numflow/graphics/geometry.py
import math
import numpy as np

def glyphVertCone(sampling=10):
    vert = []
    add = 2 * math.pi / sampling
    size = 0.5
    
    i = 0
    while i < 2 * math.pi:
        vert.extend([0, size * math.cos(i), size * math.sin(i)])
        vert.extend([0, size * math.cos(i + add), size * math.sin(i + add)])
        vert.extend([size * 2, 0, 0])
        i += add
    
    i = 0
    while i < 2 * math.pi:
        vert.extend([0, size * math.cos(i + add), size * math.sin(i + add)])
        vert.extend([0, size * math.cos(i), size * math.sin(i)])
        vert.extend([0, 0, 0])
        i += add

    return np.ascontiguousarray(np.array(vert), dtype=np.float32)


def glyphNormCone(sampling=10):
    vert = []
    add = 2 * math.pi / sampling
    t = np.array((2, 0, 0))

    i = 0
    while i < 2 * math.pi:
        a = np.array((0, math.cos(i), math.sin(i)))
        b = np.array((0, math.cos(i + add), math.sin(i + add)))
        x = a - t
        y = b - t
        c = np.cross(x, y)
        i += add

        for k in range(3):
            vert.extend([c[0], c[1], c[2]])

    i = 0
    while i < 2 * math.pi:
        for k in range(3):
            vert.extend([-1, 0, 0])
        i += add

    return np.array(vert, dtype=np.float32)


def glyphVertLine(sampling, height=4):
    vert = []
    add = 2 * math.pi / sampling 
    size = 0.5 

    i = 0
    while i < 2 * math.pi:
        vert.extend([0, size * math.cos(i), size * math.sin(i)]) 
        vert.extend([0, size * math.cos(i + add), size * math.sin(i + add)]) 
        vert.extend([size * height, size * math.cos(i), size * math.sin(i)]) 

        vert.extend([0, size * math.cos(i + add), size * math.sin(i + add)]) 
        vert.extend([size * height, size * math.cos(i + add), size * math.sin(i + add)]) 
        vert.extend([size * height, size * math.cos(i), size * math.sin(i)]) 
        i += add
    
    i = 0
    while i < 2 * math.pi:
        vert.extend([0, size * math.cos(i + add), size * math.sin(i + add)]) 
        vert.extend([0, size * math.cos(i), size * math.sin(i)]) 
        vert.extend([0, 0, 0]) 
        i += add
    
    i = 0
    while i < 2 * math.pi:
        vert.extend([size * height, size * math.cos(i), size * math.sin(i)]) 
        vert.extend([size * height, size * math.cos(i + add), size * math.sin(i + add)]) 
        vert.extend([size * height, 0, 0]) 
        i += add
    
    return vert 


def glyphNormLine(sampling, height=4):
    vert = [] 
    add = 2 * math.pi / sampling
    t = np.array((0, 0, 0))

    i = 0
    while i < 2 * math.pi:
        a = np.array((0, math.cos(i), math.sin(i)))
        b = np.array((0, math.cos(i + add), math.sin(i + add)))
        t = np.array((height, math.cos(i), math.sin(i)))
        x = a - t
        y = b - t 
        c = np.cross(x, y) 
        i += add

        for k in range(6):
            vert.extend([c[0], c[1], c[2]]) 
        
    i = 0
    while i < 2 * math.pi:
        for k in range(3):
            vert.extend([-1, 0, 0]) 
        i += add
    
    i = 0
    while i < 2 * math.pi:
        for k in range(3):
            vert.extend([1, 0, 0])
        i += add

    return vert 


def streamVert(sampling, divisions=10):
    vert = [] 

    for i in range(divisions):
        vert.extend(glyphVertLine(sampling, 0)) 

    return vert 

def streamNorm(sampling, divisions=10):
    vert = [] 

    for i in range(divisions):
        vert.extend(glyphNormLine(sampling, 1)) 
    
    return vert


def streamLocalT(sampling, divisions=10):
    vert = [] 
    add = 2 * math.pi / sampling 

    for i in range(divisions):
        base = i / divisions 
        top = (i + 1) / divisions 

        i = 0
        while i < 2 * math.pi:
            vert.extend([base, base, top, base, top, top]) 
            i += add
            
        i = 0
        while i < 2 * math.pi:
            vert.extend([base, base, base]) 
            i += add
    
        i = 0
        while i < 2 * math.pi:
            vert.extend([top, top, top]) 
            i += add
        
    return vert


def layerElements(sampling, normal):
    del sampling[normal]

    elements = [] 
    for b in range(0, int(sampling[1] - 1)): 
        for a in range(0, int(sampling[0] - 1)): 
            elements.extend([a + 1 + b * sampling[0], a + b * sampling[0], a + sampling[0] + b * sampling[0]]) 
            elements.extend([a + 1 + b * sampling[0], a + sampling[0] + b * sampling[0], a + 1 + sampling[0] + b * sampling[0]]) 

    return elements 


def colorbar(length):
    vert = [] 
    x = 1 

    for i in range(length - 1):
        vert.extend([0, i, 0]) 
        vert.extend([0, i - 1, 0]) 
        vert.extend([x, i, 0]) 

        vert.extend([x, i, 0]) 
        vert.extend([0, i - 1, 0]) 
        vert.extend([x, i - 1, 0]) 
    
    return vert 


def unitQuad():
    return [-0.5, -0.5, 0,
        0.5, -0.5, 0,
        0.5, 0.5, 0,
        -0.5, -0.5, 0,
        0.5, 0.5, 0,
        -0.5, 0.5, 0,]


def unitQuadTex():
    return [0, 1,
        1, 1,
        1, 0,
        0, 1,
        1, 0,
        0, 0,]


def generateBox():
    return np.array([
        0, 0, 0,
        0, 0, 1,

        0, 0, 1,
        0, 1, 1,

        0, 1, 1,
        0, 1, 0,

        0, 1, 0,
        0, 0, 0,

        0, 0, 0,
        1, 0, 0,

        0, 0, 1,
        1, 0, 1,

        0, 1, 1,
        1, 1, 1,

        0, 1, 0,
        1, 1, 0,

        1, 0, 0,
        1, 0, 1,

        1, 0, 1,
        1, 1, 1,

        1, 1, 1,
        1, 1, 0,

        1, 1, 0,
        1, 0, 0
    ], dtype=np.float32)